### PR TITLE
[consensus] Disable multi leader per round in universal committer

### PR DESCRIPTION
**Pull request by arun-koshy**
## Description 

On the boundaries of Leader Schedule change it is possible to elect a new leader for the same round resulting in multiple commits for the same round. This was unintended output from consensus as we have set the parameters to only commit 1 leader per round. Sui also only expects one commit per round, because of this the second commit of the same round is ignored by consensus handler. This led to an error in simtests where the `EndOfPublish` transaction was part of this ignored commit and we could not form a quorum to close the epoch without this node. 

Changes
-  Explicitly set number of leaders per round to 1 for Mysticeti commits. I wanted to try and maintain this but it would be a much more intrusive change and we will probably have to do a larger refactor to fully support it anyways, so better off doing the cleaner mitigation for now. 
- Pass back the block that the transaction was included in as part of the tx acknowledgment and include that in the logging to help with debugging future issues.
- Also a minor change to print if all tests pass during simtest seed-search 

## Test plan 

```
❯ scripts/simtest/seed-search.py simtest --test test_reconfig_with_committee_change_basic --num-seeds 1000
    Updating git repository `https://github.com/MystenLabs/mysten-sim.git`
   Compiling consensus-core v0.1.0 (/Users/arunkoshy/Documents/GitHub/sui/consensus/core)
   ...
   All tests passed successfully!
  Killing child processes
  [1]    391 killed     scripts/simtest/seed-search.py simtest --test  --num-seeds 1000
```

---

## Release notes

- [X] Protocol: Version 50 - Explicitly set number of leaders per round to 1 for Mysticeti commits
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

**Review by akichidis — APPROVED** [review]
Great find and thanks for the fix @arun-koshy ! 
Personally I would like us to have though a test that will specifically test this case - when schedule changes and different leader that the `last_decided` is elected. Since we do have the flag we can confirm that with the flag enabled/disabled we get either a double commit for same round and single with the fix. Do you believe we can add this?

File: crates/sui-core/src/mysticeti_adapter.rs [hunk 1]
```diff
@@ -95,6 +95,21 @@ impl SubmitToConsensus for LazyMysticetiClient {
...
+            && matches!(
+                transactions[0].kind,
+                ConsensusTransactionKind::EndOfPublish(_)
+                    | ConsensusTransactionKind::CapabilityNotification(_)
+                    | ConsensusTransactionKind::RandomnessDkgMessage(_, _)
+                    | ConsensusTransactionKind::RandomnessDkgConfirmation(_, _)
+            )
+        {
+            let transaction_key = SequencedConsensusTransactionKey::External(transactions[0].key());
+            tracing::info!("Transaction {transaction_key:?} was included in {block_ref}",)
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply crates/sui-core/src/mysticeti_adapter.rs:111]: I'd like to keep this, these messages are rare (I think each validator will only send ~1 of each per epoch) and being able to see them easily in logs is quite valuable.

[the remark this answers is not in the document]
Comment (akichidis) [reply crates/sui-core/src/mysticeti_adapter.rs:111]: Ok if that's the rate and info makes sense - I though some of them could be used more frequently (ex the DKG ones)

File: consensus/core/src/universal_committer.rs [hunk 2]
```diff
@@ -43,11 +43,26 @@ impl UniversalCommitter {
 
         // Try to decide as many leaders as possible, starting with the highest round.
         let mut leaders = VecDeque::new();
+
+        let last_round = if self
+            .context
+            .protocol_config
+            .mysticeti_disable_multi_leader_per_round_in_committer()
```
Comment (mwtian) [inline consensus/core/src/universal_committer.rs:50]: Can we combine this feature flag with the `NUM_LEADERS_PER_ROUND` constant?

**Review by mwtian — APPROVED** [review]

File: consensus/core/src/transaction.rs [hunk 3]
```diff
@@ -172,7 +174,7 @@ impl TransactionClient {
 
     /// Submits a list of transactions to be sequenced. The method returns when all the transactions have been successfully included
     /// to next proposed blocks.
-    pub async fn submit(&self, transactions: Vec<Vec<u8>>) -> Result<(), ClientError> {
+    pub async fn submit(&self, transactions: Vec<Vec<u8>>) -> Result<BlockRef, ClientError> {
```
[the remark this answers is not in the document]
Comment (arun-koshy) [reply consensus/core/src/transaction.rs:177]: Good point! Left a todo but wanted to ensure my understanding is correct here, if we have too many transactions for one block we will acknowledge that the transactions were included in 'Block A' and then create a new `TransactionGuard` with the remaining transactions. The remaining transactions should get included the subsequent block(s) but their acknowledgment will not be received by Sui? So like you said Sui will print out that all transactions were included in 'Block A' but may span 'Block B, C, etc.'

File: consensus/core/src/universal_committer.rs (hunk 2 above)
Comment (arun-koshy) [reply consensus/core/src/universal_committer.rs]: Good idea!!

**Review by arun-koshy — COMMENTED** [review]
Spoke offline with the team and decided to stick with the simpler solution for now and will investigate multi leader per round in the future. Also added some tests for this case

File: consensus/core/src/transaction.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (akichidis) [reply consensus/core/src/transaction.rs:177]: So on the above example the `Block A` transactions will not get acknowledged as the `ack` channel will only get passed with the last set of transactions. Let's assume that the transactions are enough to span across three blocks, `Block A, B & C` . The ack channel will only be provided to the closure that performs the acknowledgment only when the last transactions have been included to `Block C`. So the last block is the one that performs the acknowledgment

[the remark this answers is not in the document]
Comment (mwtian) [reply consensus/core/src/transaction.rs:177]: Personally I think just returning the last block is fine. When the last block is committed, blocks before should usually be committed. If we try to be precise here, we will need to return the mapping between transactions and blocks.

[the remark this answers is not in the document]
Comment (akichidis) [reply consensus/core/src/transaction.rs:177]: Yeah I guess we could be as precise/less precise want to. As I said, this only concerns soft bundles so wouldn't prioritise any time soon if needs to change.

**Pull request merged**